client-apis/streaming.md

```diff
@@ -0,0 +1,111 @@
+[//]: # "Copyright Amazon.com Inc. or its affiliates. All Rights Reserved."
+[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"
+
+# Streaming
+
+## Version
+
+0.1.0
+
+## Changelog
+
+- 0.1.0
+
+  - [Clarify Streaming Encrypt and Decrypt](../changes/2020-07-06_clarify-streaming-encrypt-decrypt/change.md)
+
+## Overview
+
+The AWS Encryption SDK MAY provide APIs that enable streamed [encryption](encrypt.md)
+and [decryption](decrypt.md).
+Streaming is a framework for making bytes available to be processed
+by an operation sequentially and over time,
+and for outputting the result of that processing
+sequentially and over time.
+
+If an implementation requires holding the entire input in memory in order to perform the operation,
+that implementation SHOULD NOT provide an API that allows the caller to stream the operation.
+
+## Definitions
+
+### Conventions used in this document
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in [RFC2119](https://tools.ietf.org/html/rfc2119).
+
+### Available Bytes
+
+In the scope of an operation, bytes are considered available if
+the operation has not yet processed those bytes
+and those bytes are intended to be processed.
+This intention is expressed through the specific streaming interface.
+
+For example, in a framework where a customer is requesting output bytes from an operation
+and that operation must read from some source in order to produce bytes,
+bytes read from the source are considered available and MUST be processed.
+Here it is the operation expressing intent to process a set of input bytes.
+Even if more bytes MAY be read from the source,
+bytes are not considered available to an operation
+until they have been consumed from the source.
+
+In a framework where a customer is sending input bytes to an operation
+and that operation must write those bytes to some sink,
+the bytes received from the customer are considered available and MUST be processed.
+Here the customer is expressing intent to process their supplied bytes.
+
+### Release
+
+An operation releases processed bytes when those bytes are outputted by the operation.
+
+If bytes are processed by an operation, that does not imply that the operation is allowed to
```

Comment:
I mean process here. i.e. you deserialized the bytes of the header, you still can't release them until the header auth is checked. "produce" seems to imply "output", which doesn't make sense here.

Reply:
True. I was more thinking "if bytes are ready to be released by the operation". Thinking more generically, a serialization operation doesn't process bytes as input, but does produce them in a stream. Since our only instances of streaming to date have streaming input AND output this isn't a major point to worry about.

Reply:
Actually, I take that back. :) I was allowing for the fact that you can refer to "If input bytes are processed, which means the operation has bytes to produce as output" which is not precisely correct in all cases. Personally I interpret "bytes are processed" as meaning the bytes are input and NOT output.

Reply:
When I say "processed bytes" I mean, the result of processing some input bytes, not the input bytes that were processed. I can see how that would be confusing here.
I have updated to mention the result of processing.